File: azulejos/solutions/correct/nlehmann.cpp
#include <algorithm>
#include <climits>
#include <cstdio>

int main() {
  int M, N, P, Q;
  scanf("%d%d%d%d", &M, &N, &P, &Q);
  int azulejos[M][N];

  for (int i = 0; i < M; ++i) {
    for (int j = 0; j < N; ++j) {
      scanf("%d", &azulejos[i][j]);
    }
  }

  int minn = INT_MAX;
  for (auto type : {0, 1}) {
    int cost = 0;
    for (int i = 0; i < M; ++i) {
      int curr = type ? 1 - (i % 2) : i % 2;
      for (int j = 0; j < N; ++j) {
        if (azulejos[i][j] != curr) {
          cost += azulejos[i][j] == 0 ? P : Q;
        }
        curr = 1 - curr;
      }
    }
    minn = std::min(minn, cost);
  }
  printf("%d\n", minn);

  return 0;
 }

File: azulejos/solutions/correct/nlehmann.java

import java.io.*;
import java.util.*;

class azulejos {
    static public void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(in.readLine());
        int M = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());
        int P = Integer.parseInt(st.nextToken());
        int Q = Integer.parseInt(st.nextToken());

        int tiles[][] = new int[M][N];
        for (int i = 0; i < M; ++i) {
            st = new StringTokenizer(in.readLine());
            for (int j = 0; j < N; ++j) {
                tiles[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int min = Integer.MAX_VALUE;
        int types[] = {0, 1};
        for (int type : types) {
            int cost = 0;
            for (int i = 0; i < M; ++i) {
                int curr = type == 0 ? i % 2 : 1 - (i % 2);
                for (int j = 0; j < N; ++j) {
                    if (curr != tiles[i][j]) {
                        cost += tiles[i][j] == 0 ? P : Q;
                    }
                    curr = 1 - curr;
                }
            }
            min = Math.min(min, cost);
        }

        System.out.println(min);
    }
}
